cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -293,76 +320,172 @@ func (s *SegmentFilter) UnmarshalJSON(p []byte) error {
 }
 
 type PaginationOptions struct {
-	Cursor *events.Cursor `json:"cursor,omitempty"`
-	Limit  uint           `json:"limit,omitempty"`
+	Cursor *db.Cursor `json:"cursor,omitempty"`
+	Limit  uint       `json:"limit,omitempty"`
 }
 
 type GetEventsResponse struct {
 	Events       []EventInfo `json:"events"`
 	LatestLedger uint32      `json:"latestLedger"`
 }
 
-type eventScanner interface {
-	Scan(eventRange events.Range, f events.ScanFunction) (uint32, error)
-}
-
 type eventsRPCHandler struct {
-	scanner      eventScanner
+	dbReader     db.EventReader
 	maxLimit     uint
 	defaultLimit uint
+	logger       *log.Entry
+	ledgerReader db.LedgerReader
+}
+
+func combineContractIDs(filters []EventFilter) ([][]byte, error) {
+	contractIDSet := set.NewSet[string](maxFiltersLimit * maxContractIDsLimit)
+
+	for _, filter := range filters {
+		for _, contractID := range filter.ContractIDs {
+			contractIDSet.Add(contractID)
+		}
+	}
+
+	contractIDs := make([][]byte, 0, len(contractIDSet))
+	for contractID := range contractIDSet {
+		id, err := strkey.Decode(strkey.VersionByteContract, contractID)
+		if err != nil {
+			return nil, fmt.Errorf("invalid contract ID %v: ", contractID)
+		}
+
+		contractIDs = append(contractIDs, id)
+	}
+	return contractIDs, nil
+}
+
+func combineEventTypes(filters []EventFilter) []int {
+	eventTypes := make(map[int]bool)
+	for _, filter := range filters {
+		for _, eventType := range filter.EventType.Keys() {
+			eventTypeXDR := getEventTypeXDRFromEventType()[eventType]
+			eventTypes[int(eventTypeXDR)] = true
+		}
+	}
+	uniqueEventTypes := make([]int, 0, len(eventTypes))
+	for eventType := range eventTypes {
```

Comment:
You aren't checking `eventType == true/false` anywhere. Should this be a Set?

Reply:
wait but why do we need that check? if keys are populated that means it is part of our result set right?

Reply:
Yeah, I'm saying it doesn't seem like you need to be using a map to bool and can use `collections.Set[int]` instead for readability